compiler/rustc_target/src/asm/xtensa.rs

```diff
@@ -0,0 +1,127 @@
+use std::fmt;
+
+use rustc_data_structures::fx::FxIndexSet;
+use rustc_span::{Symbol, sym};
+
+use super::{InlineAsmArch, InlineAsmType, ModifierInfo};
+use crate::spec::{FramePointer, RelocModel, Target};
+
+def_reg_class! {
+    Xtensa XtensaInlineAsmRegClass {
+        reg,
+        freg,
+    }
+}
+
+impl XtensaInlineAsmRegClass {
+    pub fn valid_modifiers(self, _arch: super::InlineAsmArch) -> &'static [char] {
+        &[]
+    }
+
+    pub fn suggest_class(self, _arch: InlineAsmArch, _ty: InlineAsmType) -> Option<Self> {
+        None
+    }
+
+    pub fn suggest_modifier(
+        self,
+        _arch: InlineAsmArch,
+        _ty: InlineAsmType,
+    ) -> Option<ModifierInfo> {
+        None
+    }
+
+    pub fn default_modifier(self, _arch: InlineAsmArch) -> Option<ModifierInfo> {
+        None
+    }
+
+    pub fn supported_types(
+        self,
+        _arch: InlineAsmArch,
+    ) -> &'static [(InlineAsmType, Option<Symbol>)] {
+        match self {
+            Self::reg => types! { _: I8, I16, I32; },
+            Self::freg => types! { fp: F32; },
+        }
+    }
+}
+
+fn has_fp(
+    _arch: InlineAsmArch,
+    _reloc_model: RelocModel,
+    target_features: &FxIndexSet<Symbol>,
+    _target: &Target,
+    _is_clobber: bool,
+) -> Result<(), &'static str> {
+    if target_features.contains(&sym::fp) {
+        Ok(())
+    } else {
+        Err("target does not support floating point registers")
+    }
+}
+
+fn is_frame_pointer(
+    _arch: InlineAsmArch,
+    _reloc_model: RelocModel,
+    _target_features: &FxIndexSet<Symbol>,
+    target: &Target,
+    _is_clobber: bool,
+) -> Result<(), &'static str> {
+    // The Xtensa arch doesn't require, nor use, a dedicated frame pointer register
+    // therefore if it is not force enabled, we can assume it won't be generated.
+    // If frame pointers are enabled, we cannot use the register as a general purpose one.
+    if target.options.frame_pointer != FramePointer::MayOmit {
+        Err("frame pointer register cannot be used when frame pointers are enabled")
+    } else {
+        Ok(())
+    }
+}
+
+def_regs! {
+    Xtensa XtensaInlineAsmReg XtensaInlineAsmRegClass {
+        a2: reg = ["a2"],
+        a3: reg = ["a3"],
+        a4: reg = ["a4"],
+        a5: reg = ["a5"],
+        a6: reg = ["a6"],
+        a7: reg = ["a7"],
+        a8: reg = ["a8"],
+        a9: reg = ["a9"],
+        a10: reg = ["a10"],
+        a11: reg = ["a11"],
+        a12: reg = ["a12"],
+        a13: reg = ["a13"],
+        a14: reg = ["a14"],
+        a15: reg = ["a15"] % is_frame_pointer,
+        sar: reg = ["sar"],
```

Comment:
This shouldn't be part of the `reg` class since it doesn't usually make sense for asm to select this regiser for the `reg` constraint. It should probably be in its own clobber-only register class.